### DISCUSSION TITLE
Casting a Bitwuzla term from float to bitvector

### DISCUSSION THREAD

**Discussion by daniel-raffler** [bitwuzla/bitwuzla#103] [General] [answered]
Hello,
is there are way to recast a float as a bitvector in Bitwuzla? The opposite direction is possible with Kind::FP_TO_FP_FROM_BV:
```
#include <bitwuzla/cpp/bitwuzla.h>
#include <bitwuzla/cpp/parser.h>

#include <gmp.h>

#include <string>
#include <sstream>

#include <cassert>

using namespace bitwuzla;
using namespace bitwuzla::parser;

int main() {
  Sort bvSort = mk_bv_sort(32);
  Sort fpSort = mk_fp_sort(8,24);

  Term bvTerm = mk_bv_value(bvSort, "00111101110011001100110011001101");
  Term fpTerm = mk_term(Kind::FP_TO_FP_FROM_BV, {bvTerm}, {8, 24});

  Term fpVar = mk_const(fpSort, "a");
  Term equality = mk_term(Kind::EQUAL, {fpVar, fpTerm});
  
  Options options;
  options.set(Option::PRODUCE_MODELS, 2);
  Bitwuzla bitwuzla(options);

  bitwuzla.assert_formula(equality);
  Result result = bitwuzla.check_sat();

  assert(result == Result::SAT);

  std::cout << fpVar << " = " << bitwuzla.get_value(fpVar) << std::endl;
}
```
I've tried using Kind::FP_TO_SBV:
```

int main() {
  Sort bvSort = mk_bv_sort(32);
  Sort fpSort = mk_fp_sort(8,24);

  Term rmTerm = mk_rm_value(RoundingMode::RNE);
  
  Term fpTerm = mk_fp_value(fpSort, rmTerm, "0.1");
  Term bvTerm = mk_term(Kind::FP_TO_SBV, {rmTerm, fpTerm}, {32});

  Term bvVar = mk_const(bvSort, "a");
  Term equality = mk_term(Kind::EQUAL, {bvVar, bvTerm});
  
  Options options;
  options.set(Option::PRODUCE_MODELS, 2);
  Bitwuzla bitwuzla(options);

  bitwuzla.assert_formula(equality);
  Result result = bitwuzla.check_sat();

  assert(result == Result::SAT);

  std::cout << bvVar << " = " << bitwuzla.get_value(bvVar) << std::endl;
}
```
However, this is more of a conversion and will only work if the float is already an integer number.

Is there a way to simply recast it?

Thanks,
Daniel

**Comment by aytey** [comment, marked as answer]
The normal way to do this with an SMT solver (not specially Bitwuzla) is to say that the conversion of the bitvector you want is equal to the FP you have. This is what is specified in the [floating point theory](https://smtlib.cs.uiowa.edu/theories-FloatingPoint.shtml):

```
 "There is no function for converting from (_ FloatingPoint eb sb) to the
  corresponding IEEE 754-2008 binary format, as a bit vector (_ BitVec m) with 
  m = eb + sb, because (_ NaN eb sb) has multiple, well-defined representations.
  Instead, an encoding of the kind below is recommended, where f is a term
  of sort (_ FloatingPoint eb sb):

   (declare-fun b () (_ BitVec m))
   (assert (= ((_ to_fp eb sb) b) f))
 "
```

That is, rather than converting `f` to an BV, you convert a "new" term (`m`) to a FP and assert equality with `f`.

**Comment by daniel-raffler** [comment]
Hallo Andrew,
That's quite interesting and I wasn't aware this is actually in the SMTLIB specification. We're using a method `BitvectorFormula toIeeeBitvector(FloatingPointFormula number)` in one of our JavaSMT interfaces that needs to be implemented by all solvers that support floating point formulas. MathSAT and Z3 provide  methods in their API to do so, however, in CVC4+5 and Bitwuzla the call is missing. I've now implemented the method using the work around you suggested and it seems to be working fine. Maybe this could also be adopted for the other two solvers still missing.

Thanks for your help,
Daniel

**Reply by aytey** [reply]
I would guess that MathSAT and Z3 follow the same thing as written here: https://github.com/cvc5/cvc5/issues/5537

Basically, you _can_ define a function to do that, but then you're "making guesses" for your users about what that function could do (given there are multiple possibilities).

Anyway, glad my hint helped!
